In [1]:
try:
    !pip install -U langchain-google-genai google-generativeai scikit-learn numpy
    print("Packages installed successfully")
except Exception as e:
    print(f"Error installing packages: {e}")

Packages installed successfully


In [2]:
try:
    from google.colab import userdata
    GOOGLE_API_KEY = userdata.get("Gemini_key")
    if not GOOGLE_API_KEY:
        raise ValueError("Gemini API key not found. Please set it in Colab secrets.")
    print("API key loaded successfully")
except ImportError:
    print("Error: google.colab not available. Make sure you're running in Colab.")
except Exception as e:
    print(f"Error loading API key: {e}")

API key loaded successfully


Load Embedding Model

In [4]:
try:
    from langchain_google_genai import GoogleGenerativeAIEmbeddings

    embeddings = GoogleGenerativeAIEmbeddings(
        model="models/gemini-embedding-001",
        google_api_key=GOOGLE_API_KEY
    )
    print("Embedding model initialized successfully")
except ImportError:
    print("Error: langchain_google_genai not installed. Please install it first.")
except Exception as e:
    print(f"Error initializing embeddings: {e}")

Embedding model initialized successfully


EMBEDDINGS

In [5]:
try:
    embedding = embeddings.embed_query(
        "Artificial Intelligence is transforming technology."
    )
    print("Embedding Dimension:", len(embedding))
except AttributeError:
    print("Error: embeddings object not properly initialized")
except Exception as e:
    print(f"Error generating embedding: {e}")

Embedding Dimension: 3072


In [6]:
try:
    sentences = [
        "Python is a programming language",
        "Java is used for enterprise applications",
        "Machine learning is a branch of AI",
        "Neural networks are inspired by the brain",
        "Dogs are loyal animals",
        "Cats like sleeping",
        "Football is a popular sport",
        "Cricket is played in India",
        "Databases store information",
        "SQL queries databases",
        "Operating systems manage hardware",
        "Linux is an operating system",
        "Cloud computing provides remote resources",
        "Cybersecurity protects systems",
        "Data science uses statistics",
        "Algorithms solve problems",
        "Networks connect computers",
        "Routers forward packets",
        "Web development creates websites",
        "HTML structures web pages"
    ]
    print(f"Loaded {len(sentences)} sentences")
except Exception as e:
    print(f"Error creating sentences list: {e}")

Loaded 20 sentences


In [7]:
try:
    vectors = embeddings.embed_documents(sentences)
    print("Total embeddings:", len(vectors))
    if vectors:
        print("Dimension:", len(vectors[0]))
    else:
        print("Warning: No embeddings generated")
except AttributeError:
    print("Error: embeddings object not properly initialized")
except Exception as e:
    print(f"Error generating document embeddings: {e}")

Total embeddings: 20
Dimension: 3072


In [ ]:
print(embeddings[0][:10])

[ 0.040227    0.03933245  0.0548287  -0.00986446 -0.01914819  0.01489464
 -0.03034646  0.0341908   0.00079488 -0.02408295]


SIMILARITY SEARCH

In [8]:
try:
    from sklearn.metrics.pairwise import cosine_similarity
    import numpy as np

    query = "Artificial Intelligence"
    print(f"Query: '{query}'")

    query_embedding = embeddings.embed_query(query)

    similarities = cosine_similarity(
        [query_embedding],
        vectors
    )[0]

    top_5 = np.argsort(similarities)[-5:][::-1]

    print("Top 5 Similar Sentences:\n")

    for idx in top_5:
        print(f"Sentence: {sentences[idx]}")
        print(f"Similarity: {similarities[idx]:.4f}\n")

except ImportError:
    print("Error: Required libraries not imported. Make sure scikit-learn and numpy are installed.")
except NameError:
    print("Error: Required variables (embeddings or vectors) not defined")
except Exception as e:
    print(f"Error during similarity search: {e}")

Query: 'Artificial Intelligence'
Top 5 Similar Sentences:

Sentence: Machine learning is a branch of AI
Similarity: 0.7169

Sentence: Neural networks are inspired by the brain
Similarity: 0.6730

Sentence: Algorithms solve problems
Similarity: 0.6558

Sentence: Data science uses statistics
Similarity: 0.6363

Sentence: Python is a programming language
Similarity: 0.6358



# Notebook Documentation

## Overview
This notebook demonstrates the implementation of text embeddings using Google's Gemini AI model. It covers generating embeddings for individual sentences and batches of documents, followed by similarity search using cosine similarity.

## Dependencies
- langchain-google-genai
- google-generativeai
- scikit-learn
- numpy

## Setup Requirements

### API Key Configuration
Before running this notebook, you must set up your Google Gemini API key:

1. **Google Colab Users:**
   - Click on the 🔑 Secrets icon in the left sidebar
   - Add a new secret named `Gemini_key`
   - Paste your Google Gemini API key as the value

2. **Local Environment Users:**
   - Set environment variable: `export GEMINI_API_KEY="your_api_key_here"`
   - Or modify the code to load from a `.env` file

### Getting a Gemini API Key
1. Visit [Google AI Studio](https://makersuite.google.com/app/apikey)
2. Sign in with your Google account
3. Click "Create API Key"
4. Copy the generated key

## Code Structure

### Cell 1: Install Dependencies
Installs all required Python packages including LangChain's Google GenAI integration, scikit-learn for similarity calculations, and numpy for numerical operations.

### Cell 2: Load API Key
Retrieves the Gemini API key from Colab secrets. Includes validation to ensure the key exists before proceeding.

### Cell 3: Initialize Embedding Model
Creates an instance of GoogleGenerativeAIEmbeddings with the following configuration:
- **Model**: `models/gemini-embedding-001`
- **API Key**: Loaded from environment secrets

### Cell 4: Test Single Embedding
Demonstrates generating an embedding vector for a single text query:
- Input: "Artificial Intelligence is transforming technology"
- Output: 3072-dimensional embedding vector

### Cell 5: Create Test Sentences
Defines a dataset of 20 diverse sentences covering multiple topics:
- **Technology & Programming**: Python, Java, Operating Systems, Cloud Computing
- **Artificial Intelligence**: Machine Learning, Neural Networks, Data Science
- **Sports**: Football, Cricket
- **Animals**: Dogs, Cats
- **Web Development**: HTML, Web Development

### Cell 6: Generate Document Embeddings
Processes all 20 sentences through the embedding model:
- Batch processing using `embed_documents()` method
- Outputs total count and dimension of generated embeddings

### Cell 7: Similarity Search
Performs semantic search by:
1. Generating an embedding for the query "Artificial Intelligence"
2. Computing cosine similarity between query and all document embeddings
3. Returning top 5 most similar sentences with similarity scores
